db/src/Pos/DB/Update/Logic/Global.hs

```diff
@@ -213,18 +213,22 @@ verifyBlock genesisConfig verifyAllIsKnown (ComponentBlockMain header payload) =
         -- boundaries (first slot of a new epoch) such that updates can still
         -- be adopted during OBFT.
         initialEra <- getConsensusEra
-        initialBV  <- getAdoptedBV
+        initialBV <- getAdoptedBV
+        tipHeader <- DB.getTipHeader
         case initialEra of
-            OBFT _ -> do
-                let slotId     = header ^. headerSlotL
-                    epochIndex = siEpoch slotId
-                tipHeader <- DB.getTipHeader
-                whenEpochBoundaryObft epochIndex tipHeader $ \ei -> do
-                    logDebug $ "usVerifyBlocks OBFT: We're on epoch boundary. Running processGenesisBlock"
-                    processGenesisBlock genesisConfig ei
-                getAdoptedBV
+            OBFT _ ->
+                if header ^. epochIndexL /= tipHeader ^. epochIndexL
+                    then obftProcessGenesisBlock (header ^. epochIndexL)
+                    else pure initialBV
 
-            Original -> pure initialBV
+            Original -> do
+                case tipHeader of
+                    BlockHeaderGenesis _ -> pure initialBV
+
+                    BlockHeaderMain _ ->
+                        if header ^. epochIndexL /= tipHeader ^. epochIndexL
+                            then obftProcessGenesisBlock (header ^. epochIndexL)
+                            else pure initialBV
```

Comment:
I think this is "one of the big changes" that fixes the failure-to-adopt-update issue observed on testnet 👍